### ISSUE TITLE
[tracking] Migrate pallets to FRAME umbrella crate

### ISSUE THREAD

**Issue by re-gius** [paritytech/polkadot-sdk#6504]
The goal is to stabilize the FRAME umbrella crate, so that its use within an average pallet is relatively straightforward.

We intend to achieve this goal by migrating some pallets to `polkadot-sdk-frame`, adding types and preludes to it when necessary. As a reference, you can look into #5995.

The guidelines of the umbrella crate are as follows: 

https://github.com/paritytech/polkadot-sdk/blob/06a68bec30dd56395fb8ae03e80e0fdadc682ed9/substrate/frame/src/lib.rs#L134-L150

✅ Checklist: 

- [ ] Add `polkadot-sdk-frame` to the `Cargo.toml`
- [ ] Remove all imports from `frame_support`, `sp_runtime` and similar, and replace with the appropriate `prelude::*`
- [ ] For items that are now not in scope, use the above guideline to decide if you should add them to prelude, or reside to using `deps`
- [ ] Update `Cargo.toml` to remove all unneeded dependencies. 

👉 If you want to work on this: Please check that there is not already a merge request for the pallet that you want to work on (here or in the comments below). Please pick only one or a small set of pallets - not all at once. This keeps review times low. Add a comment below to announce the pallets you're working on.

**Pallet list**

- [ ] pallet-alliance #6541 
- [x] pallet-assets-freezer #6599 
- [x] pallet-atomic-swap #6601 
- [ ] pallet-aura #6622 
- [ ] pallet-authority-discovery #6619 
- [ ] pallet-authorship #6620 
- [ ] pallet-babe #6412
- [ ] pallet-balances #6977 
- [ ] pallet-benchmarking #7196 
- [ ] pallet-bounties #7197  
- [ ] pallet-examples-split #7224 
- [ ] pallet-im-online #6736 
- [ ] pallet-indices #6737 
- [x] pallet-insecure-randomness-collective-flip #6738 
- [ ] pallet-lottery #6740 
- [x] pallet-merkle-mountain-range #7081 
- [ ] pallet-message-queue #7212 
- [ ] pallet-metadata-hash-extensions and #7202 
- [ ] pallet-migrations #7209 
- [x] pallet-mixnet #6986 
- [x] pallet-multisig #5995  
- [x] pallet-nft-fractionalization #7970 
- [x] pallet-nis #6293 
- [x] pallet-node-authorization #7040 
- [ ] pallet-nomination-pools and sub-pallets #7216 + #7218 + #7219 
- [ ] pallet-offences #7213 
- [x] pallet-paged-list #6931 
- [x] pallet-paged-list-fuzzer #6930 
- [x] pallet-proxy #5995 
- [x] pallet-recovery #7882
- [ ] pallet-remark #7228 
- [x] pallet-safe-mode #6905 
- [x] pallet-salary #7048 
- [ ] pallet-sassafras #7255 
- [ ] pallet-scored-pool and #7256 
- [ ] pallet-society #7258 
- [ ] pallet-staking-reward-curve #6412
- [ ] pallet-statement #7106 
- [ ] pallet-sudo #7106 
- [ ] pallet-timestamp #6523 
- [ ] pallet-tips #6532 
- [x] pallet-tx-pause #6630 
- [ ] pallet-verify-signature #6515 
- [x] pallet-whitelist #6514

**Comment by bennethxyz** [comment]
Can I pick this up @re-gius

**Comment by sylvaincormier** [comment]
Dear @re-gius,

I’m interested in contributing to the "Migrate pallets to umbrella crate" initiative. After reviewing the tracking issue (#6504), I’d like to start with [pallet-whitelist (#6514)], which is currently listed as open.

I’ve reviewed the guidelines and previous migration references, including pallet-multisig and pallet-proxy. I’ll ensure that the migration adheres to the umbrella crate guidelines and includes comprehensive testing.

Please let me know if this pallet is still available or if there are any considerations I should keep in mind. Looking forward to your feedback!

**Comment by runcomet** [comment]
Some updates:
pallet-alliance https://github.com/paritytech/polkadot-sdk/pull/6541
pallet-asset-conversion
pallet-asset-rate
pallet-assets-freezer
pallet-atomic-swap
pallet-aura
pallet-authority-discovery
pallet-authorship

**Comment by bennethxyz** [comment]
PRs for this Issue:

- pallet-timestamp https://github.com/paritytech/polkadot-sdk/pull/6523
- pallet-tips https://github.com/paritytech/polkadot-sdk/pull/6532
- pallet-(asset-conversion-tx-payment, asset-tx-payment, skip-feeless-payment, transaction-payment) https://github.com/paritytech/polkadot-sdk/pull/6625
- pallet-transaction-storage https://github.com/paritytech/polkadot-sdk/pull/6627
- pallet-treasury https://github.com/paritytech/polkadot-sdk/pull/6629
- pallet-tx-pause https://github.com/paritytech/polkadot-sdk/pull/6630
- pallet-uniques https://github.com/paritytech/polkadot-sdk/pull/6511
- pallet-utility https://github.com/paritytech/polkadot-sdk/pull/6512
- pallet-verify-signature https://github.com/paritytech/polkadot-sdk/pull/6515
- pallet-vesting https://github.com/paritytech/polkadot-sdk/pull/6513
- pallet-whitelist https://github.com/paritytech/polkadot-sdk/pull/6514

**Comment by kianenigma** [comment]
[bennethxyz](https://github.com/bennethxyz) [runcomet](https://github.com/runcomet) can you please read the new checklist that I put above, and make sure it is clear? Feel to ask more questions from me and @re-gius. Once clear, please update the PRs (most of them are not 100% correct).

**Comment by re-gius** [comment]
I updated the task description with the list of PRs that are ready for review and ran CI on them. Please notice that the issue is marked as `R0-silent`, so PRs should not need a prdoc (unless they do something more than the migration).

Reviews may be slower in December, but you can rest assured that they will be done.

**Comment by AurevoirXavier** [comment]
Pick up pallet-nis as https://github.com/paritytech/polkadot-sdk/pull/6293#discussion_r1843309590 suggested.

**Comment by bennethxyz** [comment]
@re-gius remove all prdocs on the open PRs

**Comment by Krayt78** [comment]
Did a list of the pallets with some AI shenanigans, mght help to know what is done/needs to be done. I havent checked if they all need the migration, just that they have a cargo.toml
Pls take the pallets 1 by 1 and dont make lots of draft PRs at once so we can all work on it together ^^ 

[x] pallet-alliance
[x] pallet-asset-conversion
[] pallet-asset-conversion-ops
[x] pallet-asset-rate
[x] pallet-assets
[x] pallet-assets-freezer
[x] pallet-atomic-swap
[x] pallet-aura
[x] pallet-authority-discovery
[x] pallet-authorship
[x] pallet-babe
[] pallet-bags-list
[] pallet-bags-list-fuzzer
[] pallet-bags-list-remote-tests
[] pallet-balances
[] pallet-beefy
[] pallet-beefy-mmr
[] pallet-benchmarking
[] pallet-benchmarking-pov
[] pallet-bounties
[] pallet-broker
[] pallet-child-bounties
[] pallet-collective
[] pallet-contracts
[] pallet-contracts-fixtures
[] pallet-contracts-fixtures-build
[] pallet-contracts-fixtures-contracts-common
[] pallet-contracts-mock-network
[] pallet-contracts-proc-macro
[] pallet-contracts-uapi
[] pallet-conviction-voting
[] pallet-core-fellowship
[] pallet-delegated-staking
[] pallet-democracy
[] pallet-election-provider-multi-phase
[] pallet-election-provider-multi-phase-test-staking-e2e
[] pallet-election-provider-support
[] pallet-election-provider-support-benchmarking
[] pallet-election-provider-support-solution-type
[] pallet-election-provider-support-solution-type-fuzzer
[] pallet-elections-phragmen
[] pallet-examples
[] pallet-examples-authorization-tx-extension
[] pallet-examples-basic
[] pallet-examples-default-config
[] pallet-examples-dev-mode
[] pallet-examples-frame-crate
[] pallet-examples-kitchensink
[] pallet-examples-multi-block-migrations
[] pallet-examples-offchain-worker
[] pallet-examples-single-block-migrations
[] pallet-examples-split
[] pallet-examples-tasks
[] pallet-executive
[] pallet-fast-unstake
[] pallet-glutton
[] pallet-grandpa
[x] pallet-identity
[x] pallet-im-online
[x] pallet-indices
[x] pallet-insecure-randomness-collective-flip
[x] pallet-lottery
[] pallet-membership
[] pallet-merkle-mountain-range
[] pallet-message-queue
[] pallet-metadata-hash-extension
[] pallet-migrations
[] pallet-mixnet
[x] pallet-multisig
[] pallet-nft-fractionalization
[] pallet-nfts
[] pallet-nfts-runtime-api
[x] pallet-nis
[] pallet-node-authorization
[] pallet-nomination-pools
[] pallet-nomination-pools-benchmarking
[] pallet-nomination-pools-fuzzer
[] pallet-nomination-pools-runtime-api
[] pallet-nomination-pools-test-delegate-stake
[] pallet-nomination-pools-test-transfer-stake
[] pallet-offences
[] pallet-offences-benchmarking
[] pallet-paged-list
[] pallet-paged-list-fuzzer
[] pallet-parameters
[] pallet-preimage
[x] pallet-proxy
[] pallet-ranked-collective
[] pallet-recovery
[] pallet-referenda
[] pallet-remark
[] pallet-revive
[] pallet-revive-fixtures
[] pallet-revive-fixtures-build
[] pallet-revive-fixtures-contracts-common
[] pallet-revive-mock-network
[] pallet-revive-proc-macro
[] pallet-revive-rpc
[] pallet-revive-uapi
[] pallet-root-offences
[] pallet-root-testing
[] pallet-safe-mode
[] pallet-salary
[] pallet-sassafras
[] pallet-scheduler
[] pallet-scored-pool
[] pallet-session
[] pallet-session-benchmarking
[] pallet-society
[] pallet-staking
[x] pallet-staking-reward-curve
[] pallet-staking-reward-fn
[] pallet-staking-runtime-api
[] pallet-state-trie-migration
[] pallet-statement
[] pallet-sudo
[] pallet-support
[] pallet-support-procedural
[] pallet-support-procedural-tools
[] pallet-support-procedural-tools-derive
[] pallet-support-test
[] pallet-support-test-compile-pass
[] pallet-support-test-pallet
[] pallet-support-test-stg-frame-crate
[] pallet-system
[] pallet-system-benchmarking
[] pallet-system-rpc-runtime-api
[x] pallet-timestamp
[x] pallet-tips
[x] pallet-transaction-payment
[x] pallet-transaction-storage
[x] pallet-treasury
[x] pallet-tx-pause
[x] pallet-uniques
[x] pallet-utility
[x] pallet-verify-signature
[x] pallet-vesting
[x] pallet-whitelist

**Comment by Krayt78** [comment]
I will start with pallet-balances

**Comment by Krayt78** [comment]
@re-gius do those that dont have pallet in their package name need to be migrated ? (ex: frame-metadata-hash-extension, frame-executive ...)

**Comment by AurevoirXavier** [comment]
Pick `pallet-balances`.

**Comment by Krayt78** [comment]
Can pallets that are still in V1 benchmarking be converted ? Or doest it need first to be converted ?

**Comment by Krayt78** [comment]
taking pallet-recovery

**Comment by Krayt78** [comment]
taking safe mode

**Comment by FasmaHouse** [comment]
why don't we add `log` & `scale-info` in the umbrella crate also? seems to be prevalent in palletes.

**Comment by FasmaHouse** [comment]
Why do we have to import `codec`, `scale-info` & `log` every-time even when it is in the umbrella crate?

**Comment by AurevoirXavier** [comment]
> Why do we have to import `codec`, `scale-info` & `log` every-time even when it is in the umbrella crate?

Because they hardcoded some of those thing in the frame-support's macros. Suggest changing those marcos usage to `frame::codec` instead of `::crate::codec`/`crate::codec`.

**Comment by FasmaHouse** [comment]
Thanks for the quick response @AurevoirXavier 

> Suggest changing those marcos usage to frame::codec instead of ::crate::codec/crate::codec.

Should we not all do this then?

**Comment by AurevoirXavier** [comment]
> Should we not all do this then?

I'm happy to see that.

**Comment by FasmaHouse** [comment]
Maybe at the final stages, when all crates are ported then one PR can help correct all the other crates?

**Comment by AurevoirXavier** [comment]
> Maybe at the final stages, when all crates are ported then one PR can help correct all the other crates?

Yes.

**Comment by Krayt78** [comment]
taking pallet scheduler

**Comment by Krayt78** [comment]
> > Why do we have to import `codec`, `scale-info` & `log` every-time even when it is in the umbrella crate?
> 
> Because they hardcoded some of those things in the frame-support's macros. Suggest changing those marcos usage to `frame::codec` instead of `::crate::codec`/`crate::codec`.

oh yeah i am having the same issue with the defensive! macro in frame support, i have no idea how to fix that without rewritting the macro to not user frame support

**Comment by re-gius** [comment]
Thanks for al your work on this PR. I should get a few reviews done by this week, and then pick up new PRs in January.

I'll ask other engineers from Parity to help, but also notice that the end of December is usually an inactivity period for low-priority tasks like this.

One important thing I want to mention is: **read the guidelines**. Don't be scared to add imports to the umbrella crate preludes, as longs as they are used across multiple pallets. If something like:

```
import A1;
import A2;
...
import A10;
```

becomes

```
import frame::{A1, A2, ..., A10);
```

then you're probably doing it wrong. The ultimate goal is to only have something like `import frame::prelude::*` and, even when that is not possible, imports should be significantly leaner than before.

**Comment by Krayt78** [comment]
Taking pallet-paged-list and pallet-paged-list-fuzzer since they go together

**Comment by seemantaggarwal** [comment]
Hi, I'd like to work on this. I will pickup `pallet-child-bounties` to start with.

**Comment by AurevoirXavier** [comment]
Pick `pallet-salary`.

**Comment by FasmaHouse** [comment]
taking `pallet-exampl-task` & `pallet-example-offchain-worker`

**Comment by tobi-michael** [comment]
taking `pallet-executive`

**Comment by Nathy-bajo** [comment]
taking `pallet-sudo` & `pallet-statement`

**Comment by tifecool** [comment]
taking `pallet-paraameter` & `pallet-preimage`

**Comment by runcomet** [comment]
Just a quick note from my recent changes if it helps anyone save time because running `cargo check` takes time. 

Overall I think it best to add systems in the prelude as there are needed in the particular pallet one is migrating and also if there exists in more than one pallet.

For example I considered adding `frame_support::traits::tokens::*;` in the top prelude but ended up causing an error with multiple `Balance` types in `docs::sdk::packages::guides::first-pallet::src::lib.rs`.

**Comment by AurevoirXavier** [comment]
> //! * The currency related traits are kept out of the preludes to encourage a deliberate choice of
//!   one over the other.

Due to the same reason. I didnot add tokens to the prelude in my PR. I'm not pretty sure how to deal with that.

**Comment by runcomet** [comment]
> > //! * The currency related traits are kept out of the preludes to encourage a deliberate choice of
> > //!   one over the other.
> 
> Due to the same reason. I didnot add tokens to the prelude in my PR. I'm not pretty sure how to deal with that.

Which one of them? @AurevoirXavier

**Comment by re-gius** [comment]
As in #6202 , we will reward each contributor with a small tip (20 DOT) once every two pallets (or one big pallet).

**Comment by re-gius** [comment]
> Why are we migrating to the `frame` crate and I see that we need to enable an experimental feature? This makes absolutely no sense. Either the crate is experimental or not. By pulling in one of these crates, the experimental feature is enabled in the entire build tree, which renders this feature completely useless.

I think that we will remove the experimental feature flag after a few merged PRs like this, when the umbrella crate will be more stable. What do you think @kianenigma ?

**Comment by re-gius** [comment]
For those of you that have open PRs, please notice that as we start merging PRs that add items to the umbrella crate preludes, you need to be careful with merging your changes since there may be overlaps.

**Comment by UtkarshBhardwaj007** [comment]
taking `pallet-node-authorization`

**Comment by Krayt78** [comment]
Opened an issue to convert the devensive! macro to be compatible with umbrella crate here #7054

**Comment by yrong** [comment]
pickup ```pallet-mmr```

**Comment by acatangiu** [comment]
> > Why are we migrating to the `frame` crate and I see that we need to enable an experimental feature? This makes absolutely no sense. Either the crate is experimental or not. By pulling in one of these crates, the experimental feature is enabled in the entire build tree, which renders this feature completely useless.
> 
> I think that we will remove the experimental feature flag after a few merged PRs like this, when the umbrella crate will be more stable. What do you think @kianenigma ?

You need to get rid of the `experimental` flag **before** merging PRs that make pallets depend on it.

**Comment by seemantaggarwal** [comment]
Create an issue to migrate Identity trait to be compatible with the umbrella create, here: #7112

**Comment by Nathy-bajo** [comment]
taking `pallet-session` & `pallet-session-benchmarking`

**Comment by gui1117** [comment]
I think it would be better not to use the `experimental` feature to gate frame crate: https://github.com/paritytech/polkadot-sdk/pull/7177

The reason is that we loose the information about which crate used the feature experimental before the migration.

I think we can consider `frame` experimental without using this feature name.

**Comment by re-gius** [comment]
New merging PRs should wait for #7177 to be merged.

**Comment by bennethxyz** [comment]
please review all open PRs @re-gius, experimental flag removed.

**Comment by acatangiu** [comment]
@kianenigma @bkchr what is the reasoning for "internal" pallets to have to depend on the "external"-facing umbrella crate, instead of only the subset of internal primitives they actually need?

Umbrella crate brings UX/DX value to external projects/crates that themselves depend on many crates within polkadot-sdk, I agree, but I don't see a strong reason for polkadot-sdk internal crates to also depend on it.

It is kinda annoying that if I need to compile or test just one pallet, I would need to compile everything under the umbrella. In practice, this only happens once, so it's not a big deal, but I still don't see the benefit to outweigh this annoyance.

**Comment by acatangiu** [comment]
And even a larger issue is that a major bump in any crate under the umbrella, major bumps the umbrella, which leads to major bumping ALL pallets depending on the umbrella.

E.g. making some `pub trait` change to something niche like `mmr-primitives` could in practice lead to only having to bump `pallet-mmr` and just that. With the umbrella thing, now all pallets get major bumped for no good reason.

See also https://github.com/paritytech/polkadot-sdk/issues/6987#issuecomment-2621083748

I now strongly disagree with this issue/initiative. We want to decouple our SDK as much as possible to be able to do better DevX, this issue works against that.

**Comment by re-gius** [comment]
> And even a larger issue is that a major bump in any crate under the umbrella, major bumps the umbrella, which leads to major bumping ALL pallets depending on the umbrella.
> See also [#6987 (comment)](https://github.com/paritytech/polkadot-sdk/issues/6987#issuecomment-2621083748)

This issue is specifically about the FRAME umbrella crate, so some of the issues you have raised should not apply.

> E.g. making some `pub trait` change to something niche like `mmr-primitives` could in practice lead to only having to bump `pallet-mmr` and just that. With the umbrella thing, now all pallets get major bumped for no good reason.

This is a good point, that is why we don't add anything to the FRAME umbrella crate, but just items that appear in three or more pallets, so that the crate is not affected by niche changes to one or two pallets.

> I now strongly disagree with this issue/initiative. We want to decouple our SDK as much as possible to be able to do better DevX, this issue works against that.

The main goal of the initiative was to migrate some pallets to the FRAME umbrella crate, so that it would become stable and ready to be used for an average pallet. To achieve this goal, we could also restrict ourselves to only migrate a smaller pre-defined list of most important pallets before we can consider the umbrella crate stable and then leave it as it is.

**Comment by re-gius** [comment]
@acatangiu I'm resuming this discussion to come to a conclusion and stabilize the frame umbrella crate. I can see a few options going forward:

1. All the concerns are gone and we go forward by migrating a lot of pallets to the umbrella crate, not necessarily every pallet.
2. Just merge other 10-15 pallets, with some large ones, so that the frame umbrella crate is stable and good enough to be used by us or external users.
3. Same as 2, but without actually migrating the pallets to the umbrella crate. There would be a "fake migration" for some main pallets just to add what's missing to the umbrella crate, but then only the resulting changes to the umbrella crate will be merged on master.

I would go with 2, is that okay with you? (also see my previous message)
We can also think of removing some import from the umbrella crate if it's too specific (like `mmr-primitives` you mentioned)

**Comment by kR1s0147** [comment]
can i pick pallet-nft-fractionalization ?

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/new-to-polkadot-best-way-to-actively-contribute-or-get-involved/12460/5

**Comment by Krayt78** [comment]
> [@acatangiu](https://github.com/acatangiu) I'm resuming this discussion to come to a conclusion and stabilize the frame umbrella crate. I can see a few options going forward:
> 
> 1. All the concerns are gone and we go forward by migrating a lot of pallets to the umbrella crate, not necessarily every pallet.
> 2. Just merge other 10-15 pallets, with some large ones, so that the frame umbrella crate is stable and good enough to be used by us or external users.
> 3. Same as 2, but without actually migrating the pallets to the umbrella crate. There would be a "fake migration" for some main pallets just to add what's missing to the umbrella crate, but then only the resulting changes to the umbrella crate will be merged on master.
> 
> I would go with 2, is that okay with you? (also see my previous message) We can also think of removing some import from the umbrella crate if it's too specific (like `mmr-primitives` you mentioned)


any wishes on which pallet you guys want first ?

**Comment by Krayt78** [comment]
taking pallet core-fellowship

**Comment by Wolfenheimm** [comment]
PR opened @ https://github.com/paritytech/polkadot-sdk/pull/8250 for pallet-grandpa 😄

**Comment by re-gius** [comment]
As far as I know, the only currently active PRs are #6931 #8264 #8285 . I hope they can be merged by the end of this week, then I will proceed by stabilizing the frame umbrella crate (whether these PRs are merged or not). 

No other PRs will be **needed** afterwards, unless there is a good reason to add the umbrella crate to a specific pallet.

**Comment by aurexav** [comment]
> As far as I know, the only currently active PRs are #6931 #8264 #8285 . I hope they can be merged by the end of this week, then I will proceed by stabilizing the frame umbrella crate (whether these PRs are merged or not). 
> 
> No other PRs will be **needed** afterwards, unless there is a good reason to add the umbrella crate to a specific pallet.

So, should I abort pallet-balances one?

**Comment by re-gius** [comment]
> > As far as I know, the only currently active PRs are [#6931](https://github.com/paritytech/polkadot-sdk/pull/6931) [#8264](https://github.com/paritytech/polkadot-sdk/pull/8264) [#8285](https://github.com/paritytech/polkadot-sdk/pull/8285) . I hope they can be merged by the end of this week, then I will proceed by stabilizing the frame umbrella crate (whether these PRs are merged or not).
> > No other PRs will be **needed** afterwards, unless there is a good reason to add the umbrella crate to a specific pallet.
> 
> So, should I abort pallet-balances one?

Unless your PR is going to be ready for review soon, I would say it's better to abort it.

**Comment by Krayt78** [comment]
> As far as I know, the only currently active PRs are [#6931](https://github.com/paritytech/polkadot-sdk/pull/6931) [#8264](https://github.com/paritytech/polkadot-sdk/pull/8264) [#8285](https://github.com/paritytech/polkadot-sdk/pull/8285) . I hope they can be merged by the end of this week, then I will proceed by stabilizing the frame umbrella crate (whether these PRs are merged or not).
> 
> No other PRs will be **needed** afterwards, unless there is a good reason to add the umbrella crate to a specific pallet.

PR #6905 is still waiting for a last reviewer
and #8285 only has that ambiguous import shenanigan and its good to go also !

**Comment by re-gius** [comment]
#8264 and #8285 are left to be fixed and possibly reviewed by the end of the week.

It seems that the FRAME umbrella crate is in a good state, pretty stable and easy to integrate in an average pallet. I'm opening a PR to stabilize it, integrating it into the main Polkadot SDK umbrella crate and adding docs.
